### PR TITLE
Checkstyle: add ArrayTypeStyle rule #319

### PR DESCRIPTION
**Pull request by eugenepeh**
#319
Ready for review

### CODE REVIEW

**Comment by CanIHasReview** [comment]
# v1

@eugenepeh submitted v1 for review.

* [v1 1/1] [Checkstyle: add ArrayTypeStyle rule](https://github.com/se-edu/addressbook-level4/pull/363/commits/74ed65b805cde654a43a6651ff9dce504e1199ac)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/27cb4621dca1bf474b2838352a90873f3fbcb4ef...74ed65b805cde654a43a6651ff9dce504e1199ac))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/363/1/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Comment by CanIHasReview** [comment]
# v2

@eugenepeh submitted v2 for review.

* [v2 1/1] [Checkstyle: add ArrayTypeStyle rule](https://github.com/se-edu/addressbook-level4/pull/363/commits/a4972ae389c97743ef8bdc1515f1b9579b27a85a)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/27cb4621dca1bf474b2838352a90873f3fbcb4ef...a4972ae389c97743ef8bdc1515f1b9579b27a85a)) ([:chart_with_upwards_trend: Interdiff between v1 and v2](https://github.com/se-edu/addressbook-level4/compare/7ebcd729cb88f77a040c300a1826b510a435c2c4~1...7ebcd729cb88f77a040c300a1826b510a435c2c4))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/363/2/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

File: config/checkstyle/checkstyle.xml [hunk 1]
```diff
@@ -154,6 +154,10 @@
     <!--
```
Comment (damithc) [inline config/checkstyle/checkstyle.xml]: > However, as Sun allows ...

Um, Sun is no longer the owner of Java. In any case, can we avoid mentioning the company?

Comment (eugenepeh) [inline config/checkstyle/checkstyle.xml]: got it

**Comment by CanIHasReview** [comment]
# v3

@eugenepeh submitted v3 for review.

* [v3 1/1] [Checkstyle: add ArrayTypeStyle rule](https://github.com/se-edu/addressbook-level4/pull/363/commits/d1726c8e588faadcba665769fd5894d348a7fa93)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/27cb4621dca1bf474b2838352a90873f3fbcb4ef...d1726c8e588faadcba665769fd5894d348a7fa93)) ([:chart_with_upwards_trend: Interdiff between v2 and v3](https://github.com/se-edu/addressbook-level4/compare/95c96e7e25cf1f540e23b1120eb8957ccdb58a15~1...95c96e7e25cf1f540e23b1120eb8957ccdb58a15))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/363/3/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

File: config/checkstyle/checkstyle.xml [hunk 2]
```diff
@@ -154,6 +154,10 @@
...
     LENGTH and CODING CHECKS
     -->
+    <module name="ArrayTypeStyle">
+      <!-- Checks that array type declarations follow Java Style -->
+      <property name="severity" value="error"/>
```
Comment (chao1995) [inline config/checkstyle/checkstyle.xml]: The `severity` property is "error", which is the same as default. By convention we don't add this line.

Comment (PierceAndy) [inline config/checkstyle/checkstyle.xml]: If that's the case, should other such lines be removed from the checkstyle file as well?

**Comment by chao1995** [comment]
About commit message,

> Array specifiers in declaration that use the C style such as
int a[] = new int[20] should not be allowed. This coding standard is
being silently ignored due to our IDE allows both C and Java style
forms of array specifiers.

We allow array declaration in both Java and C styles, e.g.
Java style: public static void main(String[] args)
C style:    public static void main(String args[]).

>This is because there are no rules in our checkstyle configuration
file that enforce this check.

However, Java style is preferred and we should only use this style
in our project.

>Let's fix this by adding ArrayTypeStyle rule. It enforces the check
on array specifiers in declaration and ensures that only the Java
style of array specifiers are allowed.

Let's add ArrayTypeStyle rule to ensure that only the Java style
of array declaration is used.

File: config/checkstyle/checkstyle.xml [hunk 3]
```diff
@@ -154,6 +154,10 @@
     <!--
     LENGTH and CODING CHECKS
     -->
+    <module name="ArrayTypeStyle">
+      <!-- Checks that array type declarations follow Java Style -->
```
Comment (chao1995) [inline config/checkstyle/checkstyle.xml]: Also, I think it's good to give examples as comment here what Java Style looks like and what C style looks like.

File: config/checkstyle/checkstyle.xml (hunk 2 above)
Comment (chao1995) [inline config/checkstyle/checkstyle.xml]: Yes we can.

Comment (PierceAndy) [inline config/checkstyle/checkstyle.xml]: Created issue for this: #370

**Comment by CanIHasReview** [comment]
# v4

@eugenepeh submitted v4 for review.

* [v4 1/1] [Checkstyle: add ArrayTypeStyle rule](https://github.com/se-edu/addressbook-level4/pull/363/commits/a734bc0dc14cfc7fdf3c158e8282d6e8904ae4f2)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/a61db1bf27f62cc549eb79be0a4598f2f1994b9b...a734bc0dc14cfc7fdf3c158e8282d6e8904ae4f2)) ([:chart_with_upwards_trend: Interdiff between v3 and v4](https://github.com/se-edu/addressbook-level4/compare/fba59ffe59efb2e928d9c615b4f300a76ec82538~1...fba59ffe59efb2e928d9c615b4f300a76ec82538))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/363/4/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Pull request merged**